README.md
golz4
=====

[![Godoc](http://img.shields.io/badge/godoc-reference-blue.svg?style=flat)](https://godoc.org/github.com/DataDog/golz4) [![license](http://img.shields.io/badge/license-BSD-red.svg?style=flat)](https://raw.githubusercontent.com/DataDog/golz4/master/LICENSE)

Golang interface to LZ4 compression.

Forked from `github.com/cloudflare/golz4` but with significant differences:

* input/output arg order has been swapped to follow Go convention, ie `Compress(in, out)` -> `Compress(out, in)`
* builds against the liblz4 that it detects using pkgconfig

Benchmark 
```
BenchmarkCompress-8             	 5000000	       234 ns/op	 183.73 MB/s	       0 B/op	       0 allocs/op
BenchmarkCompressUncompress-8   	20000000	        62.4 ns/op	 688.60 MB/s	       0 B/op	       0 allocs/op
BenchmarkStreamCompress-8       	   50000	     32842 ns/op	2003.41 MB/s	  278537 B/op	       4 allocs/op
BenchmarkStreamUncompress-8     	  500000	      2867 ns/op	22855.34 MB/s	      52 B/op	       2 allocs/op
```


CHANGELOG.md
# golz4 CHANGELOG

## v1.3.0

* Fixes the bug in Writer that was introduced in v1.2.0. [PR 19](https://github.com/DataDog/golz4/pull/19)
* Fixes a bug in how LZ4 buffers are passed to LZ4_compress_fast_continue. This bug depends on the malloc implementation. It seems rare to be triggered using glibc on Linux, but common using malloc on Mac OS X. However, this bug could happen on all systems. [PR 19](https://github.com/DataDog/golz4/pull/19), [PR 27](https://github.com/DataDog/golz4/pull/27)
* Fixes Reader to never return an negative byte count from Read. [PR 20](https://github.com/DataDog/golz4/pull/20)
* Removes an unused field from Writer and CompressReader. [PR 28](https://github.com/DataDog/golz4/pull/28)
* Updated tests to use python3. [PR 15](https://github.com/DataDog/golz4/pull/15)
* A variety of small fixes to the unit tests, including additional test coverage, and fixing the benchmarks so they run.


## v1.2.0

* *DO NOT USE*: Fixed go vet warnings about unsafe use of SliceHeader. Added a bug to Writer. [PR 17](https://github.com/DataDog/golz4/pull/17)


## v1.1.0

* Introduces new Readers, CompressReader and DecompressReader
* CompressReader provides an interface for passing an io.Reader for compression
and returning an io.ReadCloser for reading the compressed data.
* DecompressReader mirrors the functionality of the existing Reader but with
2x performance and fewer allocs. Reader is now deprecated in favor of this new type.

## v1.0.3

* Writer now supports any input size, not just blocks smaller than 65 KB. [PR 10](https://github.com/DataDog/golz4/pull/10)
* Writer ensures the double buffer used for writing do not move in memory. [PR 11](https://github.com/DataDog/golz4/pull/11)

## v1.0.2

* Fix panic with read when provided a buffer smaller than the decompressed data. The new version buffers the inflated data for later read calls when this happens. [PR 9](https://github.com/DataDog/golz4/pull/9)

## v1.0.1

Do not use deprecated LZ4 functions anymore. This removes the warnings that show up during compilation. The API or its behavior remains unchanged.

## v1.0.0

While this release **does not break API compatibility**, it changes the way the library is built.
Starting with this version, the C source code for `liblz4` **is not included in the Go package anymore**.

**The `liblz4` needs to be provided externally**, using a package manager or a manual, from source installation, for example.

Detection of `liblz4` now relies on `pkg-config` to add the correct `CFLAGS` and `LDFLAGS`.

## v0.0.131

* Initial release, using lz4 source code version r131


header_test.go
package lz4

import (
	"bytes"
	"errors"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"testing"
	"time"
)

// test extended compression/decompression w/ headers

func TestCompressHdrRatio(t *testing.T) {
	input, err := ioutil.ReadFile(sampleFilePath)
	if err != nil {
		t.Fatal(err)
	}
	output := make([]byte, CompressBoundHdr(input))
	outSize, err := CompressHdr(output, input)
	if err != nil {
		t.Fatal(err)
	}

	outputNoHdr := make([]byte, CompressBoundHdr(input))
	outNoHdrSize, err := Compress(outputNoHdr, input)
	if err != nil {
		t.Fatal(err)
	}
	if outSize != outNoHdrSize+4 {
		t.Fatalf("Compressed output length != expected: %d != %d", outSize, outNoHdrSize+4)
	}
}

func TestCompressionHdr(t *testing.T) {
	input := []byte(strings.Repeat("Hello world, this is quite something", 10))
	output := make([]byte, CompressBoundHdr(input))
	outSize, err := CompressHdr(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed, err := UncompressAllocHdr(nil, output)
	if err != nil {
		t.Fatalf("Decompression failed: %v", err)
	}
	if string(decompressed) != string(input) {
		t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
	}
}

func TestEmptyCompressionHdr(t *testing.T) {
	input := []byte("")
	output := make([]byte, CompressBoundHdr(input))
	outSize, err := CompressHdr(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed := make([]byte, len(input))
	err = UncompressHdr(decompressed, output)
	if err != nil {
		t.Fatalf("Decompression failed: %v", err)
	}
	if string(decompressed) != string(input) {
		t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
	}
}

func TestUncompressHdrShort(t *testing.T) {
	// calling Uncompress(Alloc)Hdr with input that is too short should return an error, not panic
	output := make([]byte, 1)
	for i := 0; i < 4; i++ {
		tooShortInput := make([]byte, i)
		out2, err := UncompressAllocHdr(output, tooShortInput)
		if err != errTooShort {
			t.Errorf("UncompressAllocHdr(output, [%d zero bytes]) returned unexpected err=%v",
				len(tooShortInput), err)
		}
		// UncompressAllocHdr should always returns its first argument
		// sadly slice identity is hard; cheat with Sprintf("%p")
		if fmt.Sprintf("%p", out2) != fmt.Sprintf("%p", output) {
			t.Errorf("UncompressAllocHdr([%p], [%d zero bytes]) returned output=%p",
				output, len(tooShortInput), out2)
		}

		err = UncompressHdr(output, tooShortInput)
		if err != errTooShort {
			t.Errorf("UncompressHdr(output, [%d zero bytes]) returned unexpected err=%v",
				len(tooShortInput), err)
		}
	}
}

func TestCompressAllocHdr(t *testing.T) {
	// test compressing a set of random sized inputs
	inBuf := make([]byte, 70*1024)
	for i := range inBuf {
		inBuf[i] = byte(i)
	}

	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	for i := 0; i < 1000; i++ {
		inSize := rng.Intn(len(inBuf))
		compressed, err := CompressAllocHdr(inBuf[:inSize])
		if err != nil {
			t.Fatal(err)
		}

		uncompressed, err := UncompressAllocHdr(nil, compressed)
		if err != nil {
			t.Fatal(err)
		}
		if !bytes.Equal(uncompressed, inBuf[:inSize]) {
			t.Fatal("uncompressed != input")
		}
	}
}

// test python interoperability

// pymod returns whether or not a python module is importable.  For checking
// whether or not we can test the python lz4 interop
func pymod(module string) bool {
	cmd := exec.Command("python3", "-c", fmt.Sprintf("import %s", module))
	err := cmd.Run()
	if err != nil {
		return false
	}
	return cmd.ProcessState.Success()
}

func TestPythonIntegration(t *testing.T) {
	if !pymod("os") {
		t.Errorf("pymod could not find 'os' module")
	}
	if pymod("faojfeiajwofe") {
		t.Errorf("pymod found non-existent junk module")
	}
}

func TestPythonInterop(t *testing.T) {
	pycompat := pymod("lz4.block")

	if !pycompat {
		t.Log("Warning: not testing python module compat: no module lz4 found")
		t.Skip()
		return
	}

	corpus, err := ioutil.ReadFile(sampleFilePath)
	if err != nil {
		t.Fatal(err)
	}

	out := make([]byte, CompressBoundHdr(corpus))
	count, err := CompressHdr(out, corpus)
	if err != nil {
		t.Fatal(err)
	}

	out = out[:count]

	dst := "/tmp/lz4test.z"
	err = ioutil.WriteFile(dst, out, 0644)
	if err != nil {
		t.Fatal(err)
	}
	defer os.Remove(dst)

	err = pythonLz4Compat(dst, len(corpus))
	if err != nil {
		t.Fatal(err)
	}
}

// given the original length of an lz4 encoded file, check that the python
// lz4 library returns the correct length.
func pythonLz4Compat(path string, length int) error {
	var out bytes.Buffer
	cmd := exec.Command("python3", "-c", fmt.Sprintf(`import lz4.block; print(len(lz4.block.decompress(open("%s", "rb").read())))`, path))
	cmd.Stdout = &out
	cmd.Stderr = &out
	err := cmd.Run()
	output := out.String()
	if err != nil {
		return errors.New(output)
	}
	output = strings.Trim(output, "\n")
	l, err := strconv.Atoi(output)
	if err != nil {
		return err
	}
	if l == length {
		return nil
	}
	return fmt.Errorf("Expected length %d, got %d", length, l)
}


lz4_hc_test.go
// Package lz4 implements compression using lz4.c. This is its test
// suite.
//
// Copyright (c) 2013 CloudFlare, Inc.

package lz4

import (
	"io/ioutil"
	"strings"
	"testing"
	"testing/quick"
)

func TestCompressionHCRatio(t *testing.T) {
	input, err := ioutil.ReadFile(sampleFilePath)
	if err != nil {
		t.Fatal(err)
	}
	output := make([]byte, CompressBound(input))
	outSize, err := CompressHC(output, input)
	if err != nil {
		t.Fatal(err)
	}

	// Should be at most 85% of the input size
	maxSize := 85 * len(input) / 100

	if outSize > maxSize {
		t.Fatalf("HC Compressed output length should be at most 85%% of input size: input=%d, compressed=%d, maxExpected=%d",
			len(input), outSize, maxSize)
	}
}

func TestCompressionHCLevels(t *testing.T) {
	input, err := ioutil.ReadFile(sampleFilePath)
	if err != nil {
		t.Fatal(err)
	}

	// Should be at most 85% of the input size
	previousCompressedSize := 85 * len(input) / 100

	// NOTE: lvl == 0 means auto, 1 worst, 16 best
	for lvl := 1; lvl <= 16; lvl++ {
		output := make([]byte, CompressBound(input))
		outSize, err := CompressHCLevel(output, input, lvl)
		if err != nil {
			t.Fatal(err)
		}

		if outSize > previousCompressedSize {
			t.Errorf("HC level %d should lead to a better or equal compression than HC level %d (previousSize=%d, currentSize=%d)",
				lvl, lvl-1, previousCompressedSize, outSize)
		}
		previousCompressedSize = outSize
	}
}

func TestCompressionHC(t *testing.T) {
	input := []byte(strings.Repeat("Hello world, this is quite something", 10))
	output := make([]byte, CompressBound(input))
	outSize, err := CompressHC(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed := make([]byte, len(input))
	_, err = Uncompress(decompressed, output)
	if err != nil {
		t.Fatalf("Decompression failed: %v", err)
	}
	if string(decompressed) != string(input) {
		t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
	}
}

func TestEmptyCompressionHC(t *testing.T) {
	input := []byte("")
	output := make([]byte, CompressBound(input))

	outSize, err := CompressHC(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed := make([]byte, len(input))
	_, err = Uncompress(decompressed, output)
	if err != nil {
		t.Fatalf("Decompression failed: %v", err)
	}
	if string(decompressed) != string(input) {
		t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
	}
}

func TestNoCompressionHC(t *testing.T) {
	input := []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZ")
	output := make([]byte, CompressBound(input))
	outSize, err := CompressHC(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed := make([]byte, len(input))
	_, err = Uncompress(decompressed, output)
	if err != nil {
		t.Fatalf("Decompression failed: %v", err)
	}
	if string(decompressed) != string(input) {
		t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
	}
}

func TestCompressionErrorHC(t *testing.T) {
	input := []byte(strings.Repeat("Hello world, this is quite something", 10))
	var output []byte
	outSize, err := CompressHC(output, input)

	if outSize != 0 {
		t.Fatalf("%d", outSize)
	}

	if err == nil {
		t.Fatalf("Compression should have failed but didn't")
	}

	output = make([]byte, 1)
	_, err = CompressHC(output, input)
	if err == nil {
		t.Fatalf("Compression should have failed but didn't")
	}
}

func TestDecompressionErrorHC(t *testing.T) {
	input := []byte(strings.Repeat("Hello world, this is quite something", 10))
	output := make([]byte, CompressBound(input))
	outSize, err := CompressHC(output, input)
	if err != nil {
		t.Fatalf("Compression failed: %v", err)
	}
	if outSize == 0 {
		t.Fatal("Output buffer is empty.")
	}
	output = output[:outSize]
	decompressed := make([]byte, len(input)-1)
	_, err = Uncompress(decompressed, output)
	if err == nil {
		t.Fatalf("Decompression should have failed")
	}

	decompressed = make([]byte, 1)
	_, err = Uncompress(decompressed, output)
	if err == nil {
		t.Fatalf("Decompression should have failed")
	}

	decompressed = make([]byte, 0)
	_, err = Uncompress(decompressed, output)
	if err == nil {
		t.Fatalf("Decompression should have failed")
	}
}

func TestFuzzHC(t *testing.T) {
	f := func(input []byte) bool {
		output := make([]byte, CompressBound(input))
		outSize, err := CompressHC(output, input)
		if err != nil {
			t.Fatalf("Compression failed: %v", err)
		}
		if outSize == 0 {
			t.Fatal("Output buffer is empty.")
		}
		output = output[:outSize]
		decompressed := make([]byte, len(input))
		_, err = Uncompress(decompressed, output)
		if err != nil {
			t.Fatalf("Decompression failed: %v", err)
		}
		if string(decompressed) != string(input) {
			t.Fatalf("Decompressed output != input: %q != %q", decompressed, input)
		}

		return true
	}

	conf := &quick.Config{MaxCount: 20000}
	if testing.Short() {
		conf.MaxCount = 1000
	}
	if err := quick.Check(f, conf); err != nil {
		t.Fatal(err)
	}
}


lz4_hc.go
package lz4

// #cgo pkg-config: liblz4
// #include <lz4hc.h>
import "C"

import (
	"fmt"
)

// CompressHC compresses in and puts the content in out. len(out)
// should have enough space for the compressed data (use CompressBound
// to calculate). Returns the number of bytes in the out slice. Determines
// the compression level automatically.
func CompressHC(out, in []byte) (int, error) {
	// 0 automatically sets the compression level.
	return CompressHCLevel(out, in, 0)
}

// CompressHCLevel compresses in at the given compression level and puts the
// content in out. len(out) should have enough space for the compressed data
// (use CompressBound to calculate). Returns the number of bytes in the out
// slice. To automatically choose the compression level, use 0. Otherwise, use
// any value in the inclusive range 1 (worst) through 16 (best). Most
// applications will prefer CompressHC.
func CompressHCLevel(out, in []byte, level int) (outSize int, err error) {
	// LZ4HC does not handle empty buffers. Pass through to Compress.
	if len(in) == 0 || len(out) == 0 {
		return Compress(out, in)
	}

	outSize = int(C.LZ4_compressHC2_limitedOutput(p(in), p(out), clen(in), clen(out), C.int(level)))
	if outSize == 0 {
		err = fmt.Errorf("insufficient space for compression")
	}
	return
}


lz4.go
package lz4

// #cgo pkg-config: liblz4
// #include <lz4.h>
// #include <stdlib.h>
import "C"

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"unsafe"
)

const (
	streamingBlockSize        = 1024 * 64
	blockHeaderSize           = 4
	boundedStreamingBlockSize = streamingBlockSize + streamingBlockSize/255 + 16

	hugeStreamingBlockSize        = 1024 * 1024 * 5
	boundedHugeStreamingBlockSize = hugeStreamingBlockSize + hugeStreamingBlockSize/255 + 16
)

// p gets a char pointer to the first byte of a []byte slice
func p(in []byte) *C.char {
	if len(in) == 0 {
		return (*C.char)(unsafe.Pointer(nil))
	}
	return (*C.char)(unsafe.Pointer(&in[0]))
}

// clen gets the length of a []byte slice as a char *
func clen(s []byte) C.int {
	return C.int(len(s))
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// Uncompress with a known output size. len(out) should be equal to
// the length of the uncompressed out.
func Uncompress(out, in []byte) (outSize int, err error) {
	outSize = int(C.LZ4_decompress_safe(p(in), p(out), clen(in), clen(out)))
	if outSize < 0 {
		err = errors.New("Malformed compression stream")
	}
	return
}

// CompressBound calculates the size of the output buffer needed by
// Compress. This is based on the following macro:
//
// #define LZ4_COMPRESSBOUND(isize)
//      ((unsigned int)(isize) > (unsigned int)LZ4_MAX_INPUT_SIZE ? 0 : (isize) + ((isize)/255) + 16)
func CompressBound(in []byte) int {
	return len(in) + ((len(in) / 255) + 16)
}

// Compress compresses in and puts the content in out. len(out)
// should have enough space for the compressed data (use CompressBound
// to calculate). Returns the number of bytes in the out slice.
func Compress(out, in []byte) (outSize int, err error) {
	outSize = int(C.LZ4_compress_default(p(in), p(out), clen(in), clen(out)))
	if outSize == 0 {
		err = errors.New("Insufficient space for compression")
	}
	return
}

// Writer is an io.WriteCloser that lz4 compress its input.
type Writer struct {
	compressionBuffer [2]unsafe.Pointer
	mallocBuffer      unsafe.Pointer
	lz4Stream         *C.LZ4_stream_t
	underlyingWriter  io.Writer
	inpBufIndex       int
}

// NewWriter creates a new Writer. Writes to
// the writer will be written in compressed form to w.
func NewWriter(w io.Writer) *Writer {
	// The input buffers MUST NOT be contiguous in memory. LZ4_compress_fast_continue has the
	// following comment:
	//
	// When input is structured as a double-buffer, each buffer can have any size, including < 64 KB.
	// Make sure that buffers are separated, by at least one byte.
	//
	// If the buffers are contiguous, then LZ4 assumes small writes can be "combined" for better
	// compression. For example, Write(65536); Write(1000). For the second block, it will look back
	// 64 kiB into the previous block. This means when decompressing, you will need
	// to similarly keep the buffers contiguous in memory. This code was written assuming the buffers
	// are separate. With glibc on Linux, it appears that malloc will usually allocate the buffers
	// with some space between them. However, on Mac OS X, the buffers are often contiguous.
	// See: https://github.com/lz4/lz4/issues/473#issuecomment-366537441

	// Separate the buffers so LZ4 treats them as separate. Use 8 bytes to maintain 8 byte alignment,
	// assuming malloc's result was aligned. This may permit optimizations on 64-bit CPUs.
	const bufferSeparation = 8
	mallocBuffer := C.malloc(2*streamingBlockSize + bufferSeparation)
	buffer1 := mallocBuffer
	buffer2 := unsafe.Pointer(uintptr(mallocBuffer) + streamingBlockSize + bufferSeparation)

	return &Writer{
		compressionBuffer: [2]unsafe.Pointer{buffer1, buffer2},
		mallocBuffer:      mallocBuffer,
		lz4Stream:         C.LZ4_createStream(),
		underlyingWriter:  w,
	}
}

// Write writes a compressed form of src to the underlying io.Writer.
func (w *Writer) Write(src []byte) (int, error) {
	remainingBytes := len(src)
	totalWritten := 0

	for remainingBytes > 0 {
		endIdx := totalWritten + streamingBlockSize
		if endIdx > len(src) {
			endIdx = len(src)
		}
		written, err := w.writeFrame(src[totalWritten:endIdx])
		if err != nil {
			return totalWritten, err
		}
		totalWritten += written
		remainingBytes -= written
	}

	return totalWritten, nil
}

func (w *Writer) writeFrame(src []byte) (int, error) {
	var compressedBuf [boundedStreamingBlockSize]byte
	inpPtr := w.nextInputBuffer()

	copy(inpPtr, src)

	written := int(C.LZ4_compress_fast_continue(
		w.lz4Stream,
		p(inpPtr),
		p(compressedBuf[:]),
		C.int(len(src)),
		C.int(len(compressedBuf)),
		1))
	if written <= 0 {
		return 0, errors.New("error compressing")
	}

	// Write "header" to the buffer for decompression
	var header [4]byte
	binary.LittleEndian.PutUint32(header[:], uint32(written))
	_, err := w.underlyingWriter.Write(header[:])
	if err != nil {
		return 0, err
	}

	// Write to underlying buffer
	_, err = w.underlyingWriter.Write(compressedBuf[:written])
	if err != nil {
		return 0, err
	}

	return len(src), nil
}

func (w *Writer) nextInputBuffer() []byte {
	w.inpBufIndex = (w.inpBufIndex + 1) % 2
	return unsafe.Slice((*byte)(w.compressionBuffer[w.inpBufIndex]), streamingBlockSize)
}

// Close releases all the resources occupied by Writer.
// w cannot be used after the release.
func (w *Writer) Close() error {
	if w.lz4Stream != nil {
		C.LZ4_freeStream(w.lz4Stream)
		w.lz4Stream = nil
		C.free(w.mallocBuffer)
		w.mallocBuffer = nil
	}
	return nil
}

// reader is an io.ReadCloser that decompresses when read from.
type reader struct {
	lz4Stream        *C.LZ4_streamDecode_t
	pending          []byte
	left             unsafe.Pointer
	right            unsafe.Pointer
	underlyingReader io.Reader
	isLeft           bool
}

// NewReader creates a new io.ReadCloser.  Reads from the returned ReadCloser
// read and decompress data from r.  It is the caller's responsibility to call
// Close on the ReadCloser when done.  If this is not done, underlying objects
// in the lz4 library will not be freed.
//
// Deprecated: Use NewDecompressReader instead. It can decompress the output
// of NewWriter, but uses fewer allocations.
func NewReader(r io.Reader) io.ReadCloser {
	return &reader{
		lz4Stream:        C.LZ4_createStreamDecode(),
		underlyingReader: r,
		isLeft:           true,
		// As per lz4 docs:
		//
		//   *_continue() :
		//     These decoding functions allow decompression of multiple blocks in "streaming" mode.
		//     Previously decoded blocks must still be available at the memory position where they were decoded.
		//
		// double buffer needs to use C.malloc to make sure the same memory address
		// allocate buffers in go memory will fail randomly since GC may move the memory
		left:  C.malloc(boundedStreamingBlockSize),
		right: C.malloc(boundedStreamingBlockSize),
	}
}

// Close releases all the resources occupied by r.
// r cannot be used after the release.
func (r *reader) Close() error {
	if r.lz4Stream != nil {
		C.LZ4_freeStreamDecode(r.lz4Stream)
		r.lz4Stream = nil
	}

	C.free(r.left)
	C.free(r.right)
	return nil
}

// Read decompresses `compressionBuffer` into `dst`.
// dst buffer must of at least streamingBlockSize bytes large
func (r *reader) Read(dst []byte) (int, error) {
	if len(dst) == 0 {
		return 0, nil
	}
	// Write data read from a previous call
	if r.pending != nil {
		return r.readFromPending(dst)
	}

	blockSize, err := r.readSize(r.underlyingReader)
	if err != nil {
		return 0, err
	}

	// read blockSize from r.underlyingReader --> readBuffer
	var uncompressedBuf [boundedStreamingBlockSize]byte
	_, err = io.ReadFull(r.underlyingReader, uncompressedBuf[:blockSize])
	if err != nil {
		return 0, err
	}

	var ptr unsafe.Pointer
	if r.isLeft {
		ptr = r.left
		r.isLeft = false
	} else {
		ptr = r.right
		r.isLeft = true
	}

	decompressed := int(C.LZ4_decompress_safe_continue(
		r.lz4Stream,
		(*C.char)(unsafe.Pointer(&uncompressedBuf[0])),
		(*C.char)(ptr),
		C.int(blockSize),
		C.int(streamingBlockSize),
	))

	if decompressed < 0 {
		// io.Reader requires Read to return a value in range [0, len(dst)]
		return 0, fmt.Errorf("error decompressing; result=%d", decompressed)
	}

	mySlice := C.GoBytes(ptr, C.int(decompressed))
	copySize := min(decompressed, len(dst))

	copied := copy(dst, mySlice[:copySize])

	if decompressed > len(dst) {
		// Save data for future reads
		r.pending = mySlice[copied:]
	}

	return copied, nil
}

// read the 4-byte little endian size from the head of each stream compressed block
func (r *reader) readSize(rdr io.Reader) (int, error) {
	var temp [4]byte
	_, err := io.ReadFull(rdr, temp[:])
	if err != nil {
		return 0, err
	}

	return int(binary.LittleEndian.Uint32(temp[:])), nil
}

func (r *reader) readFromPending(dst []byte) (int, error) {
	copySize := min(len(dst), len(r.pending))
	copied := copy(dst, r.pending[:copySize])

	if copied == len(r.pending) {
		r.pending = nil
	} else {
		r.pending = r.pending[copied:]
	}
	return copied, nil
}

// CompressReader reads input and creates an io.ReadCloser for reading
// compressed output
type CompressReader struct {
	underlyingReader  io.Reader
	compressionBuffer [2]unsafe.Pointer
	mallocBuffer      unsafe.Pointer
	outputBuffer      *bytes.Reader
	lz4Stream         *C.LZ4_stream_t
	inpBufIndex       int
	compressedBuffer  unsafe.Pointer
}

// NewCompressReader creates a new io.ReadCloser.  Reads from the returned ReadCloser
// read and compress data from r.  It is the caller's responsibility to call
// Close on the ReadCloser when done.  If this is not done, underlying objects
// in the lz4 library will not be freed. The compressed output must be decompressed
// using NewDecompressReader.
func NewCompressReader(r io.Reader) *CompressReader {
	// The input buffers MUST NOT be contiguous in memory so the two blocks are treated as separate.
	// We had a bug in Writer when malloc decided to allocate buffers contiguously. This bug does
	// not happen with CompressReader, because we only have "partial" blocks at EOF, and we need two
	// calls to LZ4_compress_fast_continue with sizes < 64 kiB to trigger the problem. However, we
	// should separate these buffers explicitly, to make this impossible. For details, see the
	// comment in NewWriter.

	// Separate the buffers so LZ4 treats them as separate. Use 8 bytes to maintain 8 byte alignment,
	// assuming malloc's result was aligned. This may permit optimizations on 64-bit CPUs.
	const bufferSeparation = 8
	mallocBuffer := C.malloc(2*hugeStreamingBlockSize + bufferSeparation)
	buffer1 := mallocBuffer
	buffer2 := unsafe.Pointer(uintptr(mallocBuffer) + hugeStreamingBlockSize + bufferSeparation)

	return &CompressReader{
		compressionBuffer: [2]unsafe.Pointer{buffer1, buffer2},
		mallocBuffer:      mallocBuffer,
		lz4Stream:         C.LZ4_createStream(),
		underlyingReader:  r,
		outputBuffer:      bytes.NewReader(nil),
		compressedBuffer:  C.malloc(boundedHugeStreamingBlockSize + blockHeaderSize),
	}
}

// Read compresses data from the underlyingReader into dst.
func (r *CompressReader) Read(dst []byte) (int, error) {
	// try to consume from the buffer
	n, _ := r.outputBuffer.Read(dst)
	// ignoring err which can only be EOF in which case bytes read is 0
	if n > 0 {
		// if the buffer contains anything it's leftover from a previous call
		return n, nil
	}

	// the buffer is empty, we are going to write into it so we reset it first
	totalBlockSize := boundedHugeStreamingBlockSize + blockHeaderSize
	inpPtr := r.nextInputBuffer()
	outPtr := ptrToByteSlice(r.compressedBuffer, totalBlockSize, totalBlockSize)

	bytesRead, err := io.ReadFull(r.underlyingReader, inpPtr)
	if err == io.EOF {
		// nothing left to read from the source
		return 0, err
	}
	if err != nil && err != io.ErrUnexpectedEOF {
		// ErrUnexpectedEOF occurs when some bytes are read but not all the bytes (n > 0)
		return 0, fmt.Errorf("error reading source: %s", err)
	}

	// compress and write the data into compressedBuf, leaving space for the
	// 4 byte header
	written := int(C.LZ4_compress_fast_continue(
		r.lz4Stream,
		p(inpPtr),
		p(outPtr[blockHeaderSize:]),
		C.int(bytesRead),
		C.int(boundedHugeStreamingBlockSize),
		1))
	if written <= 0 {
		return 0, errors.New("error compressing")
	}

	// write "header" to the buffer for decompression at the first 4 bytes
	binary.LittleEndian.PutUint32(outPtr[:blockHeaderSize], uint32(written))

	// populate the buffer with our internal slice and consume from it
	r.outputBuffer = bytes.NewReader(outPtr[:written+blockHeaderSize])
	n, _ = r.outputBuffer.Read(dst)
	// here we ignore any EOF because the buffer contains partial data only
	// EOF will be communicated on the next call if the underlying Reader is exhausted

	return n, nil
}

func (r *CompressReader) nextInputBuffer() []byte {
	r.inpBufIndex = (r.inpBufIndex + 1) % 2
	return ptrToByteSlice(r.compressionBuffer[r.inpBufIndex], hugeStreamingBlockSize, hugeStreamingBlockSize)
}

// Close releases all the resources occupied by Reader.
// r cannot be used after the release.
func (r *CompressReader) Close() error {
	if r.lz4Stream != nil {
		C.LZ4_freeStream(r.lz4Stream)
		r.lz4Stream = nil
		C.free(r.mallocBuffer)
		r.mallocBuffer = nil
		C.free(r.compressedBuffer)
		r.compressedBuffer = nil
	}

	return nil
}

// DecompressReader is an io.ReadCloser that decompresses when read from.
type DecompressReader struct {
	lz4Stream           *C.LZ4_streamDecode_t
	outputBuffer        *bytes.Reader
	decompressionBuffer [2]unsafe.Pointer
	underlyingReader    io.Reader
	inpBufIndex         int
	compressedBuffer    unsafe.Pointer
}

// NewDecompressReader creates a new io.ReadCloser. This function mirrors the
// behavior of NewReader but provides better performance.
// It is the caller's responsibility to call Close on the ReadCloser when done.
// If this is not done, underlying objects in the lz4 library will not be freed.
func NewDecompressReader(r io.Reader) io.ReadCloser {
	return &DecompressReader{
		lz4Stream:        C.LZ4_createStreamDecode(),
		underlyingReader: r,
		decompressionBuffer: [2]unsafe.Pointer{
			// double buffer needs to use C.malloc to make sure the same memory address
			// allocate buffers in go memory will fail randomly since GC may move the memory
			C.malloc(hugeStreamingBlockSize),
			C.malloc(hugeStreamingBlockSize),
		},
		outputBuffer:     bytes.NewReader(nil),
		compressedBuffer: C.malloc(boundedHugeStreamingBlockSize),
	}
}

// Read decompresses data from the underlying reader into `dst`.
func (r *DecompressReader) Read(dst []byte) (int, error) {
	// write data read from a previous call
	n, _ := r.outputBuffer.Read(dst)
	// ignoring err which can only be EOF in which case bytes read is 0
	if n > 0 {
		// if the buffer contains anything it's leftover from a previous call
		return n, nil
	}

	compressedBlockSize, err := r.readSize(r.underlyingReader)
	if err != nil {
		return 0, err
	}

	inPtr := ptrToByteSlice(r.compressedBuffer, boundedHugeStreamingBlockSize, boundedHugeStreamingBlockSize)
	outPtr := r.nextDecompressionBuffer()

	// read the compressed blockSize from r.underlyingReader
	_, err = io.ReadFull(r.underlyingReader, inPtr[:compressedBlockSize])
	if err != nil {
		return 0, err
	}

	decompressed := int(C.LZ4_decompress_safe_continue(
		r.lz4Stream,
		p(inPtr),
		p(outPtr),
		C.int(compressedBlockSize),
		C.int(hugeStreamingBlockSize),
	))

	if decompressed < 0 {
		return decompressed, errors.New("error decompressing")
	}

	// write the decompressed data to the output buffer
	r.outputBuffer = bytes.NewReader(outPtr[:decompressed])
	// read as much as we can into dst, ignoring any EOF
	n, _ = r.outputBuffer.Read(dst)

	return n, nil
}

// Close releases all the resources occupied by r.
// r cannot be used after the release.
func (r *DecompressReader) Close() error {
	if r.lz4Stream != nil {
		C.LZ4_freeStreamDecode(r.lz4Stream)
		r.lz4Stream = nil
	}

	C.free(r.decompressionBuffer[0])
	C.free(r.decompressionBuffer[1])
	C.free(r.compressedBuffer)
	return nil
}

func (r *DecompressReader) nextDecompressionBuffer() []byte {
	r.inpBufIndex = (r.inpBufIndex + 1) % 2
	return ptrToByteSlice(r.decompressionBuffer[r.inpBufIndex], hugeStreamingBlockSize, hugeStreamingBlockSize)
}

// read the 4-byte little endian size from the head of each stream compressed block
func (r *DecompressReader) readSize(rdr io.Reader) (int, error) {
	var temp [blockHeaderSize]byte
	_, err := io.ReadFull(rdr, temp[:])
	if err != nil {
		return 0, err
	}
	return int(binary.LittleEndian.Uint32(temp[:])), nil
}

func ptrToByteSlice(dataPtr unsafe.Pointer, _len, _cap int) []byte {
	return unsafe.Slice((*byte)(dataPtr), _len)
}
